pycaqtimage/pycaqtimage.sip

```diff
@@ -0,0 +1,727 @@
+    for (; i < (int) MAX_INDEX_PLUS1; ++i) 
+        gColorMap[i]   = u32HighValue;        
+}
+
+/*
+ * Setup an 8-bit grayscale color map.
+ */
+void pydspl_setup_gray(int iLimitLow, int iLimitHigh, int iScaleIndex)
+{
+    const uint32_t u32LowValue = ALPHA_VALUE;
+    int i = 0;
+    for (; i<= iLimitLow; ++i) 
+        gColorMap[i]   = u32LowValue;
+  
+    const int iLimitRange = iLimitHigh - iLimitLow;        
+    const float  fLimitRange = (float) iLimitRange;
+    for (; i<iLimitHigh; ++i) {
+    
+        uint8_t  u8Gray;
+        switch (iScaleIndex) {
+        case 1: /* Log2 */
+            u8Gray = (int) ( log2( 1 + (i - iLimitLow) / fLimitRange) * 255 + 0.5);
+            break;
+        case 2: /* LogE */
+            u8Gray = (int) ( log( 1 + (M_E - 1) * (i - iLimitLow) / fLimitRange) * 255 + 0.5);
+            break;
+        case 3: /* Log10 */
+            u8Gray = (int) ( log10( 1 + (10. - 1) * (i - iLimitLow) / fLimitRange) * 255 + 0.5);
+            break;
+        case 4: /* Exp2 */
+            u8Gray = (int) ( (exp2( (i - iLimitLow) / fLimitRange) - 1 ) * 255 + 0.5);
+            break;
+        case 5: /* ExpE */
+            u8Gray = (int) ( (exp( (i - iLimitLow) / fLimitRange) - 1 )/(M_E - 1) * 255 + 0.5);
+            break;
+        case 6: /* Exp10 */
+            u8Gray = (int) ( (exp10( (i - iLimitLow) / fLimitRange) - 1 )/(10. - 1) * 255 + 0.5);
+            break;
+        case 0: /* Linear */
+        default:
+            u8Gray = (i - iLimitLow) * 255 / iLimitRange;
+            break;
+        }
+          
+        uint32_t grayval  = ALPHA_VALUE | (u8Gray << 16) | (u8Gray << 8) | u8Gray;
+        gColorMap[i] = grayval;
+    }
+  
+    const uint32_t u32HighValue = ALPHA_VALUE | 0xFFFFFF;
+    for (; i < (int) MAX_INDEX_PLUS1; ++i) 
+        gColorMap[i]   = u32HighValue;    
+}
+
+/*
+ * Note: all of these are now oriented!!
+ */
+struct ImageBuffer
+{
+    QImage*   imageDisp;
+
+    uint32_t* imageData;
+    double*   projSumX;
+    double*   projSumY;
+    int       srcwidth, srcheight;
+    int       imgwidth, imgheight;
+    int       size;
+    int       iRoiW, iRoiH;
+    int       freeX, freeY, freeImage;
+  
+    int32_t  iProjXmin, iProjXmax;
+    int32_t  iProjYmin, iProjYmax;
+  
+    float fRoiPixelMean;
+    float fRoiPixelVar;
+  
+    // Frame averaging function
+    float*    imageDataF;
+    int       iAverage;
+    int       iNumAveraged;
+
+    int       isColor;
+    int       useGray;
+    int       orientation;
+};
+
+/* These must match param.py!! */
+#define ORIENT0    0
+#define ORIENT90   1
+#define ORIENT180  2
+#define ORIENT270  3
+#define ORIENT0F   4
+#define ORIENT90F  5
+#define ORIENT180F 6
+#define ORIENT270F 7
+
+static void _pyFreeImageBuffer(PyObject* pyImageBuffer)
+{
+    ImageBuffer* imageBuffer = (ImageBuffer*) PyCapsule_GetPointer(pyImageBuffer, PYC_IB);
+    
+    free(imageBuffer->imageDataF);
```

Comment:
No `if (imageBuffer->imageDataF)` guard required here?

Reply:
Not really.  The others have guards because maybe they are numpy arrays (which shouldn't be freed) and maybe they aren't (and they should).  imageDataF is always allocated in C++, so it always needs to be freed.